update-passwd.tool

```diff
@@ -0,0 +1,370 @@
+		tee >&2 <<- EOF
+			GID ${_gid} is already in use
+		EOF
+		exit 1
+	fi
+
+	echo "${_gid}"
+}
+
+
+# Usage message.
+function upUsage () {
+	tee >&2 << EOF
+usage:	update-passwd -n <Short-Name> -g <Group-Name> [-h <Home-Dir>] [-s <Shell>] -i <Info-String> -m <Group-Members>
+	update-passwd -g <Group-Name> -m <Group-Members>
+	update-passwd -V
+
+	Options include:
+	-n short-name		= User short name
+	-g group-name		= Name of (primary) group
+	-h home-dir		= Home directory path (optional)
+	-s shell		= Path to login shell (optional)
+	-i info-string		= Long discription of user
+	-m group-members	= A list of group membersips by name
+	-V			= emit version and exit
+	-?			= help message
+EOF
+	exit 1
+}
+
+# Config
+UPVERSION="&&UPVERSION&&"
+while getopts ":n:g:h:s:i:m:V" OPTION; do
+	case "${OPTION}" in
+		n)
+			SHORTNAME="${OPTARG}"
+		;;
+		g)
+			GROUPNAME="${OPTARG}"
+		;;
+		h)
+			HOME="${OPTARG}"
+		;;
+		s)
+			SHELL="${OPTARG}"
+		;;
+		i)
+			INFO="${OPTARG}"
+		;;
+		m)
+			MEMBERS="${OPTARG}"
+		;;
+		V)
+			echo "update-passwd ${UPVERSION}"
+			exit 0
+		;;
+		?)
+			# If an unknown flag is used (or -?):
+			upUsage
+		;;
+	esac
+done
+
+
+# Check if needed software is installed.
+PATH="${PATH}:/usr/local/sbin:/usr/local/bin"
+commands=(
+dscl
+defaults
+id
+sed
+grep
+cut
+tr
+tee
+)
+if [ "${sysadminctlVersionRun}" = "1" ]; then
+commands+=(
+sysadminctl
+dseditgroup
+)
+fi
+for command in "${commands[@]}"; do
+	if ! type "${command}" &> /dev/null; then
+		echo "${command} is missing, please install" >&2
+		exit 100
+	fi
+done
+
+
+### Verify that update-passwd was called correctly
+if [ "$(/usr/bin/id -u)" -ne "0" ]; then
+	echo "You must be root to run update-passwd."
+	exit 1
+fi
+
+# Set the operating mode
+if [ ! -z "${SHORTNAME}" ]; then
+	opMode="user"
+	: "${HOME:="/var/empty"}"
+	: "${SHELL:="/usr/bin/false"}"
+	if [ -z "${INFO}" ] || [ -z "${GROUPNAME}" ] || [ -z "${MEMBERS}" ]; then
+		upUsage
+	fi
+elif [ ! -z "${GROUPNAME}" ]; then
+	opMode="group"
+	if [ -z "${MEMBERS}" ]; then
+		upUsage
+	fi
+else
+	upUsage
+fi
+
+
+# Decide if ids are static or dynamic
+if [ "$(grep '^AutoUid:' "${prefixPath}/etc/passwd.conf" | sed -e 's:[[:blank:]]\{1,\}: :g' | cut -d ' ' -f "2")" = "true" ]; then
+	uidMin="$(grep '^AutoUidMin:' "${prefixPath}/etc/passwd.conf" | sed -e 's:[[:blank:]]\{1,\}: :g' | cut -d ' ' -f '2')"
+	uidMax="$(grep '^AutoUidMax:' "${prefixPath}/etc/passwd.conf" | sed -e 's:[[:blank:]]\{1,\}: :g' | cut -d ' ' -f '2')"
+elif [ -f "${prefixPath}/etc/passwd-fink.conf" ] && [ -f "${prefixPath}/etc/group-fink.conf" ]; then
+	fixedUserFink="${prefixPath}/etc/passwd-fink.conf"
+	fixedGroupFink="${prefixPath}/etc/group-fink.conf"
+fi
+
+
+# Make an array of the members list
+# membersList=($(echo "${MEMBERS}" | tr ' ' '\n'))
+
+
+
+# Setup group
+echo "Checking to see if the grooup ${GROUPNAME} exists:"
+if dscl . -read "/groups/${GROUPNAME}" PrimaryGroupID; then
```

Comment:
The `if` there is not dependent on the command output but rather the status; any status other than `0` should return false (a missing entry should return a status of 56).

Reply:
I agree and it outputs 0 on error ;)

Reply:
Interesting; if I run `dscl . -read "/groups/_redis" PrimaryGroupID; echo $?` here I get:
```
<dscl_cmd> DS Error: -14136 (eDSRecordNotFound)
56
```

and you get something like:
```
<dscl_cmd> DS Error: -14136 (eDSRecordNotFound)
0
```
? 

I hate not being able to count on exit status.

Reply:
no I get what I posted in IRC

```
dscl . -read "/groups/_redis" PrimaryGroupID
No such key: PrimaryGroupID
echo $?
0
```

Reply:
And especially interesting since I get the same output on `uname -r` = `22.5.0`, `21.6.0`, `20.6.0`, and `16.7.0`.

Reply:
yeah and this system is the same as you.

The test system is fully vanilla, nothing at all installed except XCode and XDarwin, though nothing should affect those commands that I know of. I'll do a little investigation tomorrow though it might need to wait until Thursday, but it's very odd

Reply:
kinda a far shot but could it be shell? I know on this system I switched it to bash, the other test system would be default which makes me nuts.  Trying to think of nothing, but program output shouldn't be dependant on shell :\

Reply:
`No such key: PrimaryGroupID` makes it seem like `dscl` is not working right but why/how?

Reply:
the test system is on an MDM, maybe that changes something?  But it's still something we need to plan for.  Hopefully I can figure it out.  Cause I can't see that being an issue either.

Reply:
Actually that may be the issue if it means that it is not using the local user directory.  I would be very interested in what nodes are listed in the Directory Utility app (/System/Library/CoreServices/Applications).

Reply:
@TheSin- what do you get when running `dscacheutil -configuration` on the mdm system?

Reply:
```
% sudo dscacheutil -configuration
DirectoryService Cache search policy:
    /Local/Default

Unable to get details from the cache node
Unable to get cache configuration information
```

```
% sudo dscacheutil -q group -a name clamav; echo $?
name: _clamav
password: *
gid: 82

0
```

```
% sudo dscacheutil -q group -a name _redis; echo $?
0
```

Reply:
I am going to switch over to `dscacheutil` since that seems to give consistent output across all systems. Just to double check though `id` does not give you any weirdness, right?

Reply:
dscacheutil still gives 0 when it's missing though, unless you are checking the output and not the exit code now.

Reply:
> so far id seems okay, anything specific you want me to test?

I just did it gives sane exit codes for both missing and present users.